From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Yive <6853318+Yive@users.noreply.github.com>
Date: Wed, 21 Feb 2024 07:30:25 -0800
Subject: [PATCH] Shulker configuration

Setting 'entities.shulker.disable-random-teleports' to true will also prevent splitting. It is the more efficient option if you want to disable splitting.

diff --git a/src/main/java/dev/yive/pluto/PlutoConfig.java b/src/main/java/dev/yive/pluto/PlutoConfig.java
index 174dc26cca0fad4a19c86c7c6074a2b5f65adfe8..6bd6511bcbb2da30e9f5d464e7a2327d511f94ce 100644
--- a/src/main/java/dev/yive/pluto/PlutoConfig.java
+++ b/src/main/java/dev/yive/pluto/PlutoConfig.java
@@ -380,5 +380,14 @@ public class PlutoConfig {
         private void entityDespawning() {
             entityDespawnCheckModulo = getInt("entities.global.despawn-check-modulo", entityDespawnCheckModulo);
         }
+
+        public boolean disableShulkersDroppingContentsWhenDestroyed = false;
+        public boolean disableShulkerSplitting = false;
+        public boolean disableShulkerTeleporting = false;
+        private void shulkerSettings() {
+            disableShulkersDroppingContentsWhenDestroyed = getBoolean("entities.item.disable-dropping-shulker-box-contents-when-destroyed", disableShulkersDroppingContentsWhenDestroyed);
+            disableShulkerSplitting = getBoolean("entities.shulker.disable-splitting-from-bullets", disableShulkerSplitting);
+            disableShulkerTeleporting = getBoolean("entities.shulker.disable-random-teleports", disableShulkerTeleporting);
+        }
     }
 }
diff --git a/src/main/java/net/minecraft/world/entity/monster/Shulker.java b/src/main/java/net/minecraft/world/entity/monster/Shulker.java
index 920c7a92643e83598f39bf984cca430d9deed2cd..1ca2df243e6e1c9690eb05477daf46afa8d3cf93 100644
--- a/src/main/java/net/minecraft/world/entity/monster/Shulker.java
+++ b/src/main/java/net/minecraft/world/entity/monster/Shulker.java
@@ -403,6 +403,7 @@ public class Shulker extends AbstractGolem implements VariantHolder<Optional<Dye
     }
 
     protected boolean teleportSomewhere() {
+        if (this.level().plutoConfig().disableShulkerTeleporting) return false; // Pluto - Shulker configuration
         if (!this.isNoAi() && this.isAlive()) {
             BlockPos blockposition = this.blockPosition();
 
@@ -482,6 +483,7 @@ public class Shulker extends AbstractGolem implements VariantHolder<Optional<Dye
         AABB axisalignedbb = this.getBoundingBox();
 
         if (!this.isClosed() && this.teleportSomewhere()) {
+            if (this.level().plutoConfig().disableShulkerSplitting) return; // Pluto - Shulker configuration
             int i = this.level().getEntities((EntityTypeTest) EntityType.SHULKER, axisalignedbb.inflate(8.0D), Entity::isAlive).size();
             float f = (float) (i - 1) / 5.0F;
 
diff --git a/src/main/java/net/minecraft/world/item/BlockItem.java b/src/main/java/net/minecraft/world/item/BlockItem.java
index 96fb69ec6db2e7c8c728435f0c537b076259b2fb..ef2f248ef357fb5bfb16c9298262f5b142f9a702 100644
--- a/src/main/java/net/minecraft/world/item/BlockItem.java
+++ b/src/main/java/net/minecraft/world/item/BlockItem.java
@@ -259,6 +259,7 @@ public class BlockItem extends Item {
         ItemContainerContents itemcontainercontents = (ItemContainerContents) entity.getItem().set(DataComponents.CONTAINER, ItemContainerContents.EMPTY);
 
         if (itemcontainercontents != null) {
+            if (entity.level().plutoConfig().disableShulkersDroppingContentsWhenDestroyed && this.getBlock() instanceof ShulkerBoxBlock) return; // Pluto - Shulker configuration
             ItemUtils.onContainerDestroyed(entity, itemcontainercontents.nonEmptyItemsCopy());
         }
 
